core/primitives/src/aggregate_signature.rs

```diff
@@ -154,63 +178,86 @@ impl<E: Engine> From<&SecretKey<E>> for Vec<u8> {
     }
 }
 
-// TODO(#502): TryFrom instead of From since conversion can fail
-impl<E: Engine> From<Vec<u8>> for SecretKey<E> {
-    fn from(v: Vec<u8>) -> Self {
+impl<E: Engine> FromBytes for SecretKey<E> {
+    fn from_bytes(v: Vec<u8>) -> Result<Self, Box<Error>> {
         let mut repr : <E::Fr as PrimeField>::Repr = Default::default();
         let buf = Cursor::new(v);
-        repr.read_be(buf).unwrap();
-        let scalar = <E::Fr as PrimeField>::from_repr(repr).unwrap();
-        Self{ scalar }
+        repr.read_be(buf)?;
+        let scalar = <E::Fr as PrimeField>::from_repr(repr)?;
+        Ok(Self{ scalar })
     }
 }
 
-impl<E: Engine> From<&PublicKey<E>> for Vec<u8> {
-    fn from(key: &PublicKey<E>) -> Self {
-        Self::from(key.compress().as_ref())
+impl<E: Engine> ToBytes for PublicKey<E> {
+    fn to_bytes(&self) -> Vec<u8> {
+        self.compress().as_ref().to_vec()
     }
 }
 
-// TODO(#502): TryFrom instead of From since conversion can fail
-impl<E: Engine> From<Vec<u8>> for PublicKey<E> {
-    fn from(v: Vec<u8>) -> Self {
+#[derive(Debug)]
+pub struct LengthError(usize, usize);
+
+impl Error for LengthError {
+    fn description(&self) -> &str {
+        "encoding has incorrect length"
+    }
+}
+
+impl fmt::Display for LengthError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
+        write!(f, "{}, expected {}, got {}", self.description(), self.0, self.1)
+    }
+}
+
+impl<E: Engine> FromBytes for PublicKey<E> {
+    fn from_bytes(v: Vec<u8>) -> Result<Self, Box<Error>> {
         let expected = <<E::G1Affine as CurveAffine>::Compressed as EncodedPoint>::size();
         if v.len() != expected {
-            panic!("invalid signature");
+            Err(LengthError(expected, v.len()))?;
         }
         let mut compressed = CompressedPublicKey::empty();
         compressed.as_mut().copy_from_slice(v.as_ref());
-        compressed.decompress().unwrap()
+        Ok(compressed.decompress()?)
```

Comment:
Isn't this equivalent to `compressed.decompress()`?

Reply:
No.  The Result's Err type is different.